### ISSUE TITLE
AddGeneIntegrationMatrix with 24 samples

### ISSUE THREAD

**Issue by NoemieL** [GreenleafLab/ArchR#547]
Hi,
I wondering to integrate the scRNAseq (58493 cells)to the associated scATACseq (91940 cells) data that were generated from the same sorted cells. I tried as usual (sampleCellsATAC = 10000, sampleCellsRNA = 10000) but the gene integration matrix resulted from this analysis was making nonsense since the marker to sort the cell (and other genes) has values of 0e-100. This problem was reproduced with a subset of 10 samples.
[boxplot_GeneIntegration_matrix_SDC1_TNFRSF17_clusters_samples.pdf](https://github.com/GreenleafLab/ArchR/files/6038721/boxplot_GeneIntegration_matrix_SDC1_TNFRSF17_clusters_samples.pdf)

I tried again by increasing samplesCellsATAC and sampleCallsRNA to 20000 in the subset of 10 samples, and it is looking better (ArchR-addGeneIntegrationMatrix-acbf552238715-Date-2021-02-12_Time-15-53-30.log). 
[boxplot_GeneIntegration_matrix_SDC1_TNFRSF17_clusters_samples_bis.pdf](https://github.com/GreenleafLab/ArchR/files/6038771/boxplot_GeneIntegration_matrix_SDC1_TNFRSF17_clusters_samples_bis.pdf)

However when I want to do the same in the 24 samples I get this error:

proj_all_3 <- addGeneIntegrationMatrix(ArchRProj = proj_all_2, useMatrix = "GeneScoreMatrix", matrixName = "GeneIntegrationMatrix", reducedDims = "IterativeLSI", seRNA = seRNA, sampleCellsATAC = 50000, sampleCellsRNA = 50000, nGenes = 2000, addToArrow = FALSE, groupRNA = "seurat_clusters", nameCell = "predictedCell_Un", nameGroup = "predictedGroup_Un", nameScore = "predictedScore_Un")

ArchR logging to : ArchRLogs/ArchR-addGeneIntegrationMatrix-be10e295eeb29-Date-2021-02-23_Time-18-11-18.log
If there is an issue, please report to github with logFile!
2021-02-23 18:11:18 : Running Seurat's Integration Stuart* et al 2019, 0.002 mins elapsed.
2021-02-23 18:11:23 : Checking ATAC Input, 0.074 mins elapsed.
2021-02-23 18:11:23 : Checking RNA Input, 0.074 mins elapsed.
2021-02-23 18:11:32 : Found 17211 overlapping gene names from gene scores and rna matrix!, 0.238 mins elapsed.
2021-02-23 18:11:32 : Creating Integration Blocks, 0.238 mins elapsed.
2021-02-23 18:11:36 : Prepping Interation Data, 0.295 mins elapsed.
2021-02-23 18:11:39 : Computing Integration in 2 Integration Blocks!, 0 mins elapsed.
Error in .safelapply(seq_along(blockList), function(i) { : 
Error Found Iteration 1 : 
	[1] "Error in t(x = weights) %*% prediction.mat : \n  Cholmod error 'problem too large' at file ../Core/cholmod_dense.c, line 105\n"
	<simpleError in t(x = weights) %*% prediction.mat: Cholmod error 'problem too large' at file ../Core/cholmod_dense.c, line 105>
Error Found Iteration 2 : 
	[1] "Error in t(x = weights) %*% prediction.mat : \n  Cholmod error 'problem too large' at file ../Core/cholmod_dense.c, line 105\n"
	<simpleError in t(x = weights) %*% prediction.mat: Cholmod error 'problem too large' at file ../Core/cholmod_dense.c, line 105>
In addition: Warning message:
In mclapply(..., mc.cores = threads, mc.preschedule = preschedule) :
  2 function calls resulted in an error

I also try with threads=1 and still get a error

ArchR logging to : ArchRLogs/ArchR-addGeneIntegrationMatrix-be10e6c218ebe-Date-2021-02-24_Time-10-08-38.log
If there is an issue, please report to github with logFile!
2021-02-24 10:08:39 : Running Seurat's Integration Stuart* et al 2019, 0.005 mins elapsed.
2021-02-24 10:08:42 : Checking ATAC Input, 0.067 mins elapsed.
2021-02-24 10:08:42 : Checking RNA Input, 0.067 mins elapsed.
2021-02-24 10:08:51 : Found 17211 overlapping gene names from gene scores and rna matrix!, 0.205 mins elapsed.
2021-02-24 10:08:51 : Creating Integration Blocks, 0.205 mins elapsed.
2021-02-24 10:08:54 : Prepping Interation Data, 0.258 mins elapsed.
2021-02-24 10:08:55 : Computing Integration in 2 Integration Blocks!, 0 mins elapsed.
2021-02-24 10:08:55 : Block (1 of 2) : Computing Integration, 0 mins elapsed.
2021-02-24 10:09:29 : Block (1 of 2) : Identifying Variable Genes, 0.56 mins elapsed.
2021-02-24 10:09:53 : Block (1 of 2) : Getting GeneScoreMatrix, 0.965 mins elapsed.
2021-02-24 10:11:13 : Block (1 of 2) : Imputing GeneScoreMatrix, 2.303 mins elapsed.
Getting ImputeWeights
2021-02-24 10:16:18 : Block (1 of 2) : Seurat FindTransferAnchors, 7.377 mins elapsed.
2021-02-24 12:09:53 : Block (1 of 2) : Seurat TransferData Cell Group Labels, 120.974 mins elapsed.
2021-02-24 12:10:06 : Block (1 of 2) : Seurat TransferData Cell Names Labels, 121.178 mins elapsed.
Error in t(x = weights) %*% prediction.mat : 
  Cholmod error 'problem too large' at file ../Core/cholmod_dense.c, line 105

Please, could you help me to manage this problem? also I was thinking to increase nGenes parameter since there are 17211 overlapping gene names from gene scores and rna matrix, do you think this could help to get a more accurate integration?

Thanks in advances

Noemie

**Attach your log file**
[ArchR-addGeneIntegrationMatrix-acbf552238715-Date-2021-02-12_Time-15-53-30.log](https://github.com/GreenleafLab/ArchR/files/6037651/ArchR-addGeneIntegrationMatrix-acbf552238715-Date-2021-02-12_Time-15-53-30.log)

[ArchR-addGeneIntegrationMatrix-be10e295eeb29-Date-2021-02-23_Time-18-11-18.log](https://github.com/GreenleafLab/ArchR/files/6037658/ArchR-addGeneIntegrationMatrix-be10e295eeb29-Date-2021-02-23_Time-18-11-18.log)

[ArchR-addGeneIntegrationMatrix-be10e6c218ebe-Date-2021-02-24_Time-10-08-38.log](https://github.com/GreenleafLab/ArchR/files/6038691/ArchR-addGeneIntegrationMatrix-be10e6c218ebe-Date-2021-02-24_Time-10-08-38.log)

**Comment by jgranja24** [comment]
Hi @NoemieL, the bug you are experiencing seems abnormal. Let me add some more logging and push a new version of the function to better diagnose why all your values are 0. If you increase sampleCells to 50,000 for both you will exhaust your memory especially running in parallel. I would consider doing sampleCellsATAC = 10,000 and adjusting sampleCellsRNA independently (linear memory increase) vs at the same time (exponential memory increase). However the first thing you showed seems like a bug... I would also try with threads = 1 to see if this is a memory error as well. I will update hopefully within a day for a better logging version!

**Comment by rcorces** [comment]
In case it helps...

> Cholmod error 'problem too large' 

Google seems to indicate that this error can happen when you try to convert a very large sparse matrix to dense format.

**Comment by NoemieL** [comment]
> Hi @NoemieL, the bug you are experiencing seems abnormal. Let me add some more logging and push a new version of the function to better diagnose why all your values are 0. If you increase sampleCells to 50,000 for both you will exhaust your memory especially running in parallel. I would consider doing sampleCellsATAC = 10,000 and adjusting sampleCellsRNA independently (linear memory increase) vs at the same time (exponential memory increase). However the first thing you showed seems like a bug... I would also try with threads = 1 to see if this is a memory error as well. I will update hopefully within a day for a better logging version!

**Comment by rcorces** [comment]
@NoemieL - could you let us know what solved the error so that others who encounter this can benefit?

**Comment by NoemieL** [comment]
Sorry I did a mistake and accidentally closed it....
Thanks, please let me know when you have a better logging version and I will redo it with sampleCellsATAC = 10,000.

**Comment by NoemieL** [comment]
Hi,
I reinstall ArchR with devtools::install_github("GreenleafLab/ArchR", ref="master", repos = BiocManager::repositories()) and get this Note: possible error in '.loadUWOT(embedding$params$uwotModel, ': unused argument (embedding$params$nc) but I don't think this is the problem.

I rerun the 24 samples from the beginning and redo the scRNAseq integration:

proj_all_3 <- addGeneIntegrationMatrix(
  ArchRProj = proj_all_2, 
  useMatrix = "GeneScoreMatrix",
  matrixName = "GeneIntegrationMatrix",
  reducedDims = "IterativeLSI",
  seRNA = seRNA,
  sampleCellsATAC = 10000,
  sampleCellsRNA = 10000,
  nGenes = 2000,
  addToArrow = TRUE,
  groupRNA = "seurat_clusters",
  nameCell = "predictedCell_Un",
  nameGroup = "predictedGroup_Un",
  nameScore = "predictedScore_Un"
)

I still have 0 values, with 12 threads:
[ArchR-addGeneIntegrationMatrix-c8bcb58a80d27-Date-2021-03-02_Time-14-02-20.log](https://github.com/GreenleafLab/ArchR/files/6071983/ArchR-addGeneIntegrationMatrix-c8bcb58a80d27-Date-2021-03-02_Time-14-02-20.log)

or with 1 thread:
[ArchR-addGeneIntegrationMatrix-c8bcb2efda48a-Date-2021-03-02_Time-14-50-05.log](https://github.com/GreenleafLab/ArchR/files/6072607/ArchR-addGeneIntegrationMatrix-c8bcb2efda48a-Date-2021-03-02_Time-14-50-05.log)


Hope this could help to identify the problem. Let me know if I could do somethig else.

Thanks

**Comment by NoemieL** [comment]
Hi,
I think the problem came during the normalization of the integrating RNA matrix (scaleTo).

I could run without problem 21 samples (excluding 2 samples because the mean read by cell was double compared to the other samples, cf attached info where excluding samples are highlighted in bold)
[ArchR-addGeneIntegrationMatrix-d9c8920b9aa27-Date-2021-03-15_Time-10-09-21.log](https://github.com/GreenleafLab/ArchR/files/6199024/ArchR-addGeneIntegrationMatrix-d9c8920b9aa27-Date-2021-03-15_Time-10-09-21.log)

Nb_cells | Mean_Reads_per_Cell | Median_Genes_per_Cell | Nb_Reads | Valid_Barcodes | Total_Genes_Detected
5,028 | 54,120 | 2,367 | 272,116,355 | 98.10% | 19,322
6,207 | 57,220 | 2,282 | 355,170,473 | 97.50% | 19,781
4,462 | 53,550 | 1,594 | 238,940,490 | 97.90% | 18,793
5,787 | 49,358 | 1,446 | 285,636,114 | 97.50% | 19,088
**1,152 | 117,532 | 702 | 135,397,755 | 98.20% | 14,977**
9,792 | 39,927 | 1,479 | 390,972,464 | 97.50% | 19,406
1,109 | 62,806 | 1,359 | 69,652,167 | 97.90% | 17,210
1,843 | 41,989 | 1,331 | 77,386,122 | 97.80% | 18,772
6,480 | 49,825 | 1,313 | 322,871,046 | 97.50% | 20,356
8,751 | 43,232 | 2,056 | 378,327,205 | 97.00% | 21,749
1,947 | 61,371 | 1,114 | 119,490,025 | 97.80% | 19,243
4,890 | 42,921 | 1,825 | 209,884,515 | 97.50% | 20,690
6,048 | 49,009 | 2,391 | 296,411,254 | 98.20% | 22,087
3,805 | 56,083 | 2,337 | 213,396,986 | 98.20% | 21,879
4,439 | 52,552 | 2,770 | 233,278,537 | 98.10% | 21,449
3,066 | 54,521 | 2,362 | 167,161,605 | 98.30% | 20,365
**1,986 | 109,171 | 1,037 | 216,814,129 | 98.40% | 18,511**
5,112 | 51,878 | 2,751 | 265,200,806 | 98.30% | 22,146
7,138 | 48,574 | 1,896 | 346,722,551 | 98.20% | 23,012
2,834 | 50,576 | 2,645 | 143,333,572 | 98.40% | 20,126
3,451 | 49,674 | 2,842 | 171,427,952 | 98.40% | 19,398
4,746 | 51,366 | 2,717 | 243,785,350 | 98.20% | 21,687
1,243 | 80,420 | 1,778 | 99,962,866 | 98.20% | 16,954


Then I run this 2 excluding samples together and I have not problem either. 
[ArchR-addGeneIntegrationMatrix-e8b2d4afdaa66-Date-2021-03-23_Time-16-14-51.log](https://github.com/GreenleafLab/ArchR/files/6199006/ArchR-addGeneIntegrationMatrix-e8b2d4afdaa66-Date-2021-03-23_Time-16-14-51.log)

What can I do to include these two samples in my analysis? Thanks

**Comment by rcorces** [comment]
I'm not 100% sure I follow everything here. you have two samples where the RNA libraries have been sequenced ~2x higher and you think this these two are causing problems in the integration?

when you remove both of these "deep" samples from the integration, it works as expected?

and when you run those two samples by themselves, it also works as expected?

What happens if you only remove one of them?

In general, this will be very difficult to debug without a reproducible example for us to look at.

**Comment by NoemieL** [comment]
Hi, Did you change anything? It looks like it is working well now, however I did not change anything in my script. I did remove one of the two excluded samples and it is working. I redid with all the samples and it is working too. Also I did not have to change anymore the addArchRThreads(threads = 1) for the constraining integration or final integration (which was necessary before to perform these steps).
[ArchR-addGeneIntegrationMatrix-fd8b625c9aa10-Date-2021-04-07_Time-09-01-57.log](https://github.com/GreenleafLab/ArchR/files/6273388/ArchR-addGeneIntegrationMatrix-fd8b625c9aa10-Date-2021-04-07_Time-09-01-57.log)
